pkg/proxy/iptables/proxier.go

```diff
@@ -175,7 +175,7 @@ func newServiceInfo(serviceName proxy.ServicePortName, port *api.ServicePort, se
 
 	if info.onlyNodeLocalEndpoints {
 		p := apiservice.GetServiceHealthCheckNodePort(service)
-		if p == 0 {
+		if p == 0 && service.Spec.Type == api.ServiceTypeLoadBalancer {
```

Comment:
Yeah, we should fix NeedsHealthCheck and call it here. I refined this logic in the first commit of #41162.

I don't mind if you'd like to fix NeedsHealthCheck earlier as that PR isn't totally ready. But be aware of you might also need to modified the callers of NeedsHealthCheck (like pkg/registry/core/service/rest.go).